py/reporter_service.py

```diff
@@ -96,55 +118,104 @@ def parse_trace(self, post):
     trace = self.parse_trace(post)
 
-    #lets get the uuid from json the request
+    #uuid is required
     uuid = trace.get('uuid')
-    if uuid is not None:
-      #do we already know something about this vehicleId already? Let's check Redis
-      partial = thread_local.cache.get(uuid)
-      if partial:
-        partial = pickle.loads(partial)
-        time_diff = trace['trace'][0]['time'] - partial[-1]['time']
-        #check to make sure time is not stale and not in future
-        if time_diff < os.environ.get('STALE_TIME', 60) and time_diff >= 0:
-          #Now prepend the last bit of shape from the partial_end segment that's already in Redis
-          #to the rest of the partial_start segment once it is returned from the segment_matcher
-          trace['trace'] = partial + trace['trace']
-    else:
-      return 400, 'No uuid in segment_match request!'
+    if uuid is None:
+      return 400, 'uuid is required'
 
-    #ask valhalla to give back OSMLR segments along this trace
-    result = thread_local.segment_matcher.Match(json.dumps(trace, separators=(',', ':')))
-    segments = json.loads(result)
+    #one or more points is required
+    try:
+      sq_distance, elapsed_time = sum_difference(trace['trace'])
+      trace['sq_distance'] = sq_distance
+      trace['elapsed_time'] = elapsed_time
+    except Exception as e:
+      return 400, 'trace must be a non zero length array of object eacho of which must have at least lat, lon and time'
 
-    #if there are segments
-    if len(segments['segments']):
-      #if the last one had the beginning of the ots but not the end we'll want to continue it
-      if segments['segments'][-1]['start_time'] >= 0 and segments['segments'][-1]['end_time'] < 0:
-        #gets the begin index of the last partial
-        begin_index = segments['segments'][-1]['begin_shape_index']
-        #in Redis, set the uuid as key and trace from the begin index to the end
-        thread_local.cache.set(uuid, pickle.dumps(trace['trace'][begin_index:]), ex=os.environ.get('PARTIAL_EXPIRY', 300))
-      #if any others are partial, we do not need so remove them
-      segments['segments'] = [ seg for seg in segments['segments'] if seg['length'] > 0 ]
-      segments['mode'] = "auto"
-      segments['provider'] = os.environ.get('PROVIDER', '')
-      #segments['reporter_id'] = os.environ['REPORTER_ID']
-
-      #Now we will send the whole segments on to the datastore
-      if len(segments['segments']):
-        response = requests.post(os.environ['DATASTORE_URL'], json.dumps(segments))
-        if response.status_code != 200:
-          sys.stderr.write(response.text)
-          sys.stderr.flush()
-          return 500, response.text
-
-    #hand it back
-    return 200, 'Reported on %d segments' % len(segments['segments'])
+    #do we already have some previous trace to prepend for this uuid
+    partial_segments = None
+    #TODO: aquire lock on uuid through redis, otherwise race conditions will ensue
```

Comment:
as it stands we cant merge this until we do this `TODO`

this is critical if we want to actually accept single point traces and still scale the reporter horizontally. the issue here is that if one thread/proc/machine/container gets a single point for a given `uuid` and goes to update what it has in redis, it will race with potentially another thread/proc/machine/container to do that update. what we need to prevent that is a distributed locking mechanism which we can rip off from here: https://github.com/SPSCommerce/redlock-py

you may note that this problem was always there but was somewhat mitigated that we didnt ever just do a cache operation without a match and some purging before.

we must also worry about out of order stuff, there is a check for it but somewhat assumes sequential handling...